Modules/Loadable/Volumes/Logic/vtkSlicerVolumesLogic.cxx

```diff
@@ -1545,3 +1553,128 @@ ::ResampleVolumeToReferenceVolume(vtkMRMLVolumeNode* inputVolumeNode,
 
   return outputVolumeNode;
 }
+
+
+// --------------------------------------------------------------------------
+void vtkSlicerVolumesLogic::setColorNode(vtkMRMLScalarVolumeNode* volumeNode, vtkMRMLNode* colorNode)
+{
+
+    if (!volumeNode)
+    {
+        QString message = QString("No volume node.");
+        qWarning() << Q_FUNC_INFO << ": " << message;
+        return;
+    }
+    vtkMRMLScalarVolumeDisplayNode* displayNode = volumeNode->GetScalarVolumeDisplayNode();
+    if (!displayNode)
+    {
+        QString message = QString("No display node.");
+        qWarning() << Q_FUNC_INFO << ": " << message;
+        return;
+    }
+    if (!displayNode || !colorNode)
+    {
+        QString message = QString("No display and color node.");
+        qWarning() << Q_FUNC_INFO << ": " << message;
+        return;
+    }
+    Q_ASSERT(vtkMRMLColorNode::SafeDownCast(colorNode));
+    displayNode->SetAndObserveColorNodeID(colorNode->GetID());
+}
+
+//----------------------------------------------------------------------------
+bool vtkSlicerVolumesLogic::SetVolumeWindowLevel(vtkMRMLScalarVolumeNode* volumeNode, double window, double level, bool isAutoWindowLevel)
+{
+    vtkMRMLScene* scene = this->GetMRMLScene();
+    if (!volumeNode)
+    {
+        QString message = QString("No volume node.");
+        qWarning() << Q_FUNC_INFO << ": " << message;
+        return false;
+    }
+    vtkMRMLScalarVolumeDisplayNode* volumeDisplayNode = volumeNode->GetScalarVolumeDisplayNode();
+    if (!volumeDisplayNode)
+    {
+        QString message = QString("No display node.");
+        qWarning() << Q_FUNC_INFO << ": " << message;
+        return false;
+    }
+    if (isAutoWindowLevel)
+    {
+        int disabledModify = volumeDisplayNode->StartModify();
+        volumeDisplayNode->SetWindowLevel(window, level);
+        volumeDisplayNode->SetAutoWindowLevel(1);
+        volumeDisplayNode->EndModify(disabledModify);
+    }
+    else
+    {
+        int disabledModify = volumeDisplayNode->StartModify();
+        volumeDisplayNode->SetAutoWindowLevel(0);
+        volumeDisplayNode->SetWindowLevel(window, level);
+        volumeDisplayNode->EndModify(disabledModify);
+    }
+    return true;
+}
+
+// --------------------------------------------------------------------------
+void vtkSlicerVolumesLogic::setWindowLevelPreset(vtkMRMLScalarVolumeNode* volumeNode, const QString& presetName)
```

Comment:
Yes, the display node definitely should be able to store some volume-specific presets, for example the preset(s) imported from the DICOM file.

It would be tempting to store application-wide presets in the display node, too, but that would not be consistent with how we store almost all other presets/constants in module logics (color tables, volume rendering presets, terminologies, and units). Slice view orientation preset management is an exception (those are stored in the scene), and it is not a very clean or robust design. For example, if we want to change orientation presets then we need to iterate through all existing slice nodes and modify all of them, which is somewhat complicated and it also removes all customizations that other modules may have applied. You would probably also need to update the window/level presets in all volume display nodes when you load a scene, which would not be nice.

View context menu is managed by Subject Hierarchy plugins, which can access any module logic class. See for example [here](https://github.com/Slicer/Slicer/blob/7eed7c54ea41b2a5e3c4b73c8e5d84391876e71f/Modules/Loadable/Markups/SubjectHierarchyPlugins/qSlicerSubjectHierarchyMarkupsPlugin.cxx#L291), so it should be no problem to maintain a shared list of window/level presets there.

Reply:
I wouldn't suggest saving the preset values in the as instance variables in display nodes (that is, they shouldn't be saved in the scene), only that any hard-coded values for certain displayable types should be in the display nodes and that any helper classes to manipulated them should be in the logics.  Keeping the options in the display node code also makes sense since they can then easily be augmented or overwritten by subclasses.

Reply:
How do you envision it would work? Would the data reader determine what kind of image is loaded (CT, US, MRI, PET,...custom image types) and initialize the presets in the display node?

If we don't save the presets in the scene then how would the presets would be set in display nodes that are loaded from a scene?

Reply:
What I mean is that currently these hard-coded preset defaults are [stored in the widget code](https://github.com/Slicer/Slicer/blob/550f0334a3538609e4748ec18ec77bfaddda7950/Modules/Loadable/Volumes/Widgets/qSlicerScalarVolumeDisplayWidget.cxx#L406-L476_), which is clearly the wrong place.  I think it would be cleaner to have this information provided by the display node (as class-static data).  This would be in addition to the existing window level preset list, which is in an instance variable and is saved with the scene.  Then if we did make, for example, a `vtkMRMLPETVolumeNode` with a `vtkMRMLPETVolumeDisplayNode` the display node could provide only the presets that make sense for the specialized subclass.

Reply:
How would it work for CT? Would you create a new vtkMRMLCTVolumeDisplayNode and use that instead of vtkMRMLScalarVolumeDisplayNode when you detect that it is a CT?

An image modality detection mechanism would help in recommending volume rendering presets and color node, too. This logic is currently implemented in the volume rendering subject hierarchy plugin, but it would make sense to implement it at lower level (MRML and/or module logic).

In the future, modality could be used to recommend hanging protocol and analysis modules.

Reply:
We have this pattern already for DiffusionWeighted and DiffusionTensor volumes, where they override certain behaviors in the slice views and UI.  But it's incomplete since some of the Volumes UI is hard coded.

So yes, I can imagine having a specialized volumes for modalities that expose custom rendering options.  We already do some of this when reading from dicom.  When we read from nrrd or other formats this information is lost, but we could offer the user options to specify it when known (like we do when reading a nifti as a segmentation for example).